src/processors/chain/ProcessorChain.cpp

```diff
@@ -0,0 +1,200 @@
+#include "ProcessorChain.h"
+#include "../ParameterHelpers.h"
+#include "ProcessorChainActionHelper.h"
+#include "ProcessorChainStateHelper.h"
+#include "processors/chain/ChainIOProcessor.h"
+
+namespace
+{
+[[maybe_unused]] void printBufferLevels (const AudioBuffer<float>& buffer)
+{
+    const auto numChannels = buffer.getNumChannels();
+    const auto numSamples = buffer.getNumSamples();
+    for (int ch = 0; ch < numChannels; ++ch)
+    {
+        auto level = buffer.getRMSLevel (ch, 0, numSamples);
+        std::cout << "Level for channel: " << ch << ": " << level << std::endl;
+    }
+}
+} // namespace
+
+ProcessorChain::ProcessorChain (ProcessorStore& store,
+                                AudioProcessorValueTreeState& vts,
+                                std::unique_ptr<chowdsp::PresetManager>& presetMgr) : procStore (store),
+                                                                                      um (vts.undoManager),
+                                                                                      inputProcessor (um),
+                                                                                      outputProcessor (um),
+                                                                                      ioProcessor (vts),
+                                                                                      presetManager (presetMgr)
+{
+    actionHelper = std::make_unique<ProcessorChainActionHelper> (*this);
+    stateHelper = std::make_unique<ProcessorChainStateHelper> (*this);
+
+    using namespace std::placeholders;
+    procStore.addProcessorCallback = std::bind (&ProcessorChainActionHelper::addProcessor, actionHelper.get(), _1);
+    procStore.replaceProcessorCallback = std::bind (&ProcessorChainActionHelper::replaceProcessor, actionHelper.get(), _1, _2);
```

Comment:
This should probably be done in the `actionHelper` constructor? And maybe we can use lambda's instead of binding...